### PR TITLE
EEP 52: Allow pins inside map keys in matches

### PR DESCRIPTION
**Pull request by v0idpwn**
Related to #9603

### CODE REVIEW

**Comment by v0idpwn** [comment]
It is failing on CI because the main branch isn't formatted

**Comment by michallepicki** [comment]
Should there also be a test that executes such code, or is expansion test enough?

**Comment by v0idpwn** [comment]
I'm working on a test soon!

**Comment by josevalim** [comment]
Also please rebase, I messed up the formatting. :)

File: lib/elixir/src/elixir_map.erl [hunk 1]
```diff
@@ -60,10 +60,10 @@ clean_struct_key_from_map_assocs(Meta, Assocs, E) ->
       Assocs
   end.
 
-validate_match_key(Meta, {'^', _, [{Name, _, Context}]}, E) when is_atom(Name), is_atom(Context) ->
-  form_error(Meta, E, ?MODULE, {invalid_pin_in_map_key_match, Name});
 validate_match_key(Meta, {Name, _, Context}, E) when is_atom(Name), is_atom(Context) ->
   form_error(Meta, E, ?MODULE, {invalid_variable_in_map_key_match, Name});
+validate_match_key(_, {'^', _, [{Name, _, Context}]}, _) when is_atom(Name), is_atom(Context) ->
```
Comment (josevalim) [inline lib/elixir/src/elixir_map.erl:65]: You can now remove a clause from `validate_kv` too.

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by v0idpwn** [comment]
Just as a note, this doesn't allow guard expressions in map match keys as in EEP52, it justs extends Elixir what Elixir supports in map match keys through EEP52.

If the syntax for complete support to this EEP52 feature is figured, I can give it a shot, but I think it probably doesn't fit very well with Elixir. 🙂

**Pull request merged**